presto-base-arrow-flight/src/main/java/com/facebook/plugin/arrow/AbstractArrowMetadata.java

```diff
@@ -0,0 +1,199 @@
+            columns = (List<ArrowColumnHandle>) (List<?>) arrowColumns;
+        }
+
+        ConnectorTableLayout layout = new ConnectorTableLayout(new ArrowTableLayoutHandle(tableHandle, columns, constraint.getSummary()));
+        return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
+    }
+
+    @Override
+    public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTableLayoutHandle handle)
+    {
+        return new ConnectorTableLayout(handle);
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        List<ColumnMetadata> meta = new ArrayList<>();
+        List<Field> columnList = getColumnsList(((ArrowTableHandle) table).getSchema(), ((ArrowTableHandle) table).getTable(), session);
+
+        for (Field field : columnList) {
+            String columnName = field.getName();
+            Type fieldType = getPrestoTypeFromArrowField(field);
+            meta.add(new ColumnMetadata(columnName, fieldType));
+        }
+        return new ConnectorTableMetadata(new SchemaTableName(((ArrowTableHandle) table).getSchema(), ((ArrowTableHandle) table).getTable()), meta);
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        return ((ArrowColumnHandle) columnHandle).getColumnMetadata();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        requireNonNull(prefix, "prefix is null");
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        List<SchemaTableName> tables;
+        if (prefix.getSchemaName() != null && prefix.getTableName() != null) {
+            tables = ImmutableList.of(new SchemaTableName(prefix.getSchemaName(), prefix.getTableName()));
+        }
+        else {
+            tables = listTables(session, Optional.of(prefix.getSchemaName()));
+        }
+
+        for (SchemaTableName tableName : tables) {
+            try {
+                ConnectorTableHandle tableHandle = getTableHandle(session, tableName);
+                columns.put(tableName, getTableMetadata(session, tableHandle).getColumns());
+            }
+            catch (ClassCastException | NotFoundException e) {
+                throw new ArrowException(ARROW_FLIGHT_METADATA_ERROR, "The table columns could not be listed for the table " + tableName, e);
+            }
+            catch (Exception e) {
+                throw new ArrowException(ARROW_FLIGHT_METADATA_ERROR, e.getMessage(), e);
+            }
+        }
+        return columns.build();
+    }
+
+    private Type getPrestoTypeFromArrowField(Field field)
+    {
+        return arrowBlockBuilder.getPrestoTypeFromArrowField(field);
```

Comment:
@tdcmeehan We are injecting an instance of `ArrowBlockBuilder` to re-use the arrow to presto field mappings.